near-sdk/src/utils/mod.rs

```diff
@@ -40,6 +40,96 @@ macro_rules! log {
     };
 }
 
+/// Helper macro to create assertions that will panic through the runtime host functions.
+///
+/// This macro can be used similarly to [`assert!`] but will reduce code size by not including
+/// file and rust specific data in the panic message.
+///
+/// # Examples
+///
+/// ```no_run
+/// use near_sdk::require;
+///
+/// # fn main() {
+/// let a = 2;
+/// require!(a > 0);
+/// require!("test" != "other", "Some custom error message if false");
+/// # }
+/// ```
+#[macro_export]
+macro_rules! require {
+    ($cond:expr $(,)?) => {
+        if !$cond {
+            $crate::env::panic_str("require! assertion failed")
+        }
+    };
+    ($cond:expr, $message:expr $(,)?) => {
+        if !$cond {
+            $crate::env::panic_str($message.as_ref())
+        }
+    };
+}
+
+/// Asserts that two expressions are equal to each other using [`PartialEq`].
+///
+/// Like [`require!`], this will panic through runtime host functions and reduce contract size with
+/// a more concise panic message.
+///
+/// # Examples
+///
+/// ```
+/// use near_sdk::require_eq;
+///
+/// # fn main() {
+/// let a = 2;
+/// require_eq!(a, 2);
+/// require_eq!("test", "test", "Some custom error message if neq");
+/// # }
+/// ```
+#[macro_export]
+macro_rules! require_eq {
+    ($left:expr, $right:expr $(,)?) => {
+        if $left != $right {
+            $crate::env::panic_str("require_eq! assertion failed")
+        }
+    };
+    ($left:expr, $right:expr, $message:expr $(,)?) => {
+        if $left != $right {
+            $crate::env::panic_str($message.as_ref())
```

Comment:
Not entirely sure, but it seems that built-in format uses `&$message` rather than `$message.as_ref()`

* https://github.com/rust-lang/rust/blob/d59b80d588368cdcfcc1d54e119374a3d78169ff/compiler/rustc_builtin_macros/src/format.rs#L797
* https://github.com/rust-lang/rust/blob/d59b80d588368cdcfcc1d54e119374a3d78169ff/compiler/rustc_expand/src/build.rs#L232-L235
* https://github.com/rust-lang/rust/blob/d59b80d588368cdcfcc1d54e119374a3d78169ff/compiler/rustc_ast/src/ast.rs#L1362-L1364

Reply:
So the rationale was that I wanted the ability to supply something that wasn't just a static string or if keeping `expr` then forcing the user to pass in a `&str`, in place of not having the formatting machinery.

I was thinking it would be used as the `log!` macro where it allows `String`, `&str`, or anything that implements `AsRef<str>`. I suppose all we want here is to allow owned or str references, so I think I will take this suggestion because it's unlikely someone will assert with the message being a single struct